### PR TITLE
Added the Selection of cells till Top and Bottom

### PR DESCRIPTION
**Pull request by bw-space**
## References

This Pull Request solves the issue: #6783

## Code changes

The following files were changed:
1. jupyterlab/packages/notebook-extension/src/index.ts
in which two new command definitions i.e extendBottom and extendTop have been added.
2. jupyterlab/packages/notebook-extension/schema/tracker.json
in which two new shortcut keys are defined i.e Shift+PageUp and Shift+PageDown to select to top and bottom respectively.
3. jupyterlab/packages/notebook/src/actions.tsx
in which the function definitions extendSelectionAbove and extendSelectionBelow have been added an extra parameter to identify selections till top or only one cell above and vice versa, respectively.

### CODE REVIEW

File: examples/notebook/src/commands.ts [hunk 1]
```diff
@@ -164,11 +166,19 @@ export const SetupCommands = (
   });
   commands.addCommand(cmdIds.extendAbove, {
     label: 'Extend Above',
-    execute: () => NotebookActions.extendSelectionAbove(nbWidget.content)
+    execute: () => NotebookActions.extendSelectionAbove(nbWidget.content, false)
+  });
+  commands.addCommand(cmdIds.extendTop, {
+    label: 'Extend Above',
```
Comment (jasongrout) [inline examples/notebook/src/commands.ts]: Can you change the label to Extend to Top, and similarly for the other new command?

File: packages/notebook/src/actions.tsx [hunk 2]
```diff
@@ -772,7 +772,7 @@ export namespace NotebookActions {
     const state = Private.getState(notebook);
 
     notebook.mode = 'command';
-    //Check if toBottom is true, if yes selection is made till bottom.
+    // Check if toBottom is true, if yes selection is made till bottom.
```
Comment (jasongrout) [inline packages/notebook/src/actions.tsx]: Can you document the new parameter above in the docstring, below the line `@param notebook ...`

**Comment by jasongrout** [comment]
How about the new toTop and toBottom parameters default to false? Then the current code does not need to change, and it is backwards compatible so this can go in for 1.2.

**Comment by jasongrout** [comment]
Thanks! I left a few review comments.

File: examples/notebook/src/commands.ts (hunk 1 above)
Comment (bw-space) [reply examples/notebook/src/commands.ts]: Sure.

File: packages/notebook/src/actions.tsx (hunk 2 above)
Comment (bw-space) [reply packages/notebook/src/actions.tsx]: Sure.

File: tests/test-notebook/src/actions.spec.ts [hunk 3]
```diff
@@ -912,79 +912,79 @@ describe('@jupyterlab/notebook', () => {
     describe('#extendSelectionAbove()', () => {
       it('should extend the selection to the cell above', () => {
         widget.activeCellIndex = 1;
-        NotebookActions.extendSelectionAbove(widget);
+        NotebookActions.extendSelectionAbove(widget, false);
```
Comment (jasongrout) [inline tests/test-notebook/src/actions.spec.ts]: With the default being false, now we don't need to change any of the existing calls in this file.

File: tests/test-notebook/src/actions.spec.ts [hunk 4]
```diff
@@ -912,79 +912,79 @@ describe('@jupyterlab/notebook', () => {
...
         expect(widget.activeCellIndex).to.equal(0);
       });
     });
 
     describe('#extendSelectionBelow()', () => {
       it('should extend the selection to the cell below', () => {
-        NotebookActions.extendSelectionBelow(widget);
+        NotebookActions.extendSelectionBelow(widget, false);
```
Comment (jasongrout) [inline tests/test-notebook/src/actions.spec.ts]: Can you add a test for the second parameter being true here and for extendSelectionAbove?

Comment (bw-space) [reply tests/test-notebook/src/actions.spec.ts]: Sure.

File: packages/notebook-extension/src/index.ts [hunk 5]
```diff
@@ -1459,7 +1463,18 @@ function addCommands(
       const current = getCurrent(args);
 
       if (current) {
-        return NotebookActions.extendSelectionAbove(current.content);
+        return NotebookActions.extendSelectionAbove(current.content, false);
```
Comment (jasongrout) [inline packages/notebook-extension/src/index.ts]: `false` is now default, so not needed

File: packages/notebook-extension/src/index.ts [hunk 6]
```diff
@@ -1470,7 +1485,18 @@ function addCommands(
       const current = getCurrent(args);
 
       if (current) {
-        return NotebookActions.extendSelectionBelow(current.content);
+        return NotebookActions.extendSelectionBelow(current.content, false);
```
Comment (jasongrout) [inline packages/notebook-extension/src/index.ts]: `false` is now default, so not needed

File: packages/notebook/src/actions.tsx [hunk 7]
```diff
@@ -719,12 +719,15 @@ export namespace NotebookActions {
    * Extend the selection to the cell above.
    *
    * @param notebook - The target notebook widget.
-   *
+   * @param toTop - If true, denotes selection to extend till the top.
```
Comment (jasongrout) [inline packages/notebook/src/actions.tsx]: Add an extra line after this, before the heading.

Comment (jasongrout) [inline packages/notebook/src/actions.tsx]:
```suggestion
   * @param toTop - If true, denotes selection to extend to the top.
```

File: packages/notebook-extension/src/index.ts (hunk 5 above)
Comment (bw-space) [reply packages/notebook-extension/src/index.ts]: Ok, I have to change it to the previous one right?

File: packages/notebook/src/actions.tsx [hunk 8]
```diff
@@ -736,20 +739,28 @@ export namespace NotebookActions {
     const state = Private.getState(notebook);
 
     notebook.mode = 'command';
-    notebook.extendContiguousSelectionTo(notebook.activeCellIndex - 1);
+    // Check if toTop is true, if yes, selection is made till top.
```
Comment (jasongrout) [inline packages/notebook/src/actions.tsx]:
```suggestion
    // Check if toTop is true, if yes, selection is made to the top.
```

File: packages/notebook/src/actions.tsx [hunk 9]
```diff
@@ -736,20 +739,28 @@ export namespace NotebookActions {
...
+    }
     Private.handleState(notebook, state, true);
   }
 
   /**
    * Extend the selection to the cell below.
    *
    * @param notebook - The target notebook widget.
-   *
+   * @param toBottom - If true, denotes selection to extend till the bottom.
```
Comment (jasongrout) [inline packages/notebook/src/actions.tsx]:
```suggestion
   * @param toBottom - If true, denotes selection to extend to the bottom.

```

File: packages/notebook/src/actions.tsx [hunk 10]
```diff
@@ -761,7 +772,12 @@ export namespace NotebookActions {
     const state = Private.getState(notebook);
 
     notebook.mode = 'command';
-    notebook.extendContiguousSelectionTo(notebook.activeCellIndex + 1);
+    // Check if toBottom is true, if yes selection is made till bottom.
```
Comment (jasongrout) [inline packages/notebook/src/actions.tsx]:
```suggestion
    // Check if toBottom is true, if yes selection is made to the bottom.
```

**Comment by jasongrout** [comment]
Thanks! I added a few more comments inline.

File: packages/notebook-extension/src/index.ts (hunk 5 above)
Comment (jasongrout) [reply packages/notebook-extension/src/index.ts]: Yes, I would just leave the previous code alone, since it should continue to work just fine.

Comment (jasongrout) [inline packages/notebook-extension/src/index.ts]:
```suggestion
        return NotebookActions.extendSelectionAbove(current.content);
```

Comment (bw-space) [reply packages/notebook-extension/src/index.ts]: I have updated the files with the above changes.

File: examples/notebook/src/commands.ts [hunk 11]
```diff
@@ -164,11 +166,19 @@ export const SetupCommands = (
   });
   commands.addCommand(cmdIds.extendAbove, {
     label: 'Extend Above',
-    execute: () => NotebookActions.extendSelectionAbove(nbWidget.content)
+    execute: () => NotebookActions.extendSelectionAbove(nbWidget.content, false)
```
Comment (jasongrout) [inline examples/notebook/src/commands.ts]:
```suggestion
    execute: () => NotebookActions.extendSelectionAbove(nbWidget.content)
```

File: examples/notebook/src/commands.ts [hunk 12]
```diff
@@ -164,11 +166,19 @@ export const SetupCommands = (
...
+  });
+  commands.addCommand(cmdIds.extendTop, {
+    label: 'Extend to Top',
+    execute: () => NotebookActions.extendSelectionAbove(nbWidget.content, true)
   });
   commands.addCommand(cmdIds.extendBelow, {
     label: 'Extend Below',
-    execute: () => NotebookActions.extendSelectionBelow(nbWidget.content)
+    execute: () => NotebookActions.extendSelectionBelow(nbWidget.content, false)
```
Comment (jasongrout) [inline examples/notebook/src/commands.ts]:
```suggestion
    execute: () => NotebookActions.extendSelectionBelow(nbWidget.content)
```

Comment (bw-space) [reply examples/notebook/src/commands.ts]: I have updated the file with the above changes.

**Comment by bw-space** [comment]
@jasongrout do I need to change any more files for accepting the pull request?

**Comment by jasongrout** [comment]
Can you look at the two outstanding comments above?

https://github.com/jupyterlab/jupyterlab/pull/7177#pullrequestreview-285833741

https://github.com/jupyterlab/jupyterlab/pull/7177#discussion_r322493606

**Comment by bw-space** [comment]
I have made the necessary changes, @jasongrout is the file ok?

File: tests/test-notebook/src/actions.spec.ts [hunk 13]
```diff
@@ -916,6 +916,12 @@ describe('@jupyterlab/notebook', () => {
         expect(widget.isSelected(widget.widgets[0])).to.equal(true);
       });
 
+      it('should extend the selection to the topmost cell', () => {
+        widget.activeCellIndex = 1;
+        NotebookActions.extendSelectionAbove(widget, true);
+        expect(widget.isSelected(widget.widgets[0])).to.equal(true);
```
Comment (jasongrout) [inline tests/test-notebook/src/actions.spec.ts]: Thanks! I think we're almost there!

This test goes to the second cell, then asks to select to the top, and checks that the one cell above it is selected. But that doesn't distinguish between the normal extendSelectionAbove behavior and extending to the very top. Can you go down to the bottom cell, and then check to see if every cell above it is selected? That would be testing that it actually is doing what it says, different than the behavior before.

File: tests/test-notebook/src/actions.spec.ts [hunk 14]
```diff
@@ -968,6 +974,11 @@ describe('@jupyterlab/notebook', () => {
         expect(widget.isSelected(widget.widgets[1])).to.equal(true);
       });
 
+      it('should extend the selection the bottomost cell', () => {
+        NotebookActions.extendSelectionBelow(widget, true);
+        expect(widget.isSelected(widget.widgets[0])).to.equal(true);
+        expect(widget.isSelected(widget.widgets[1])).to.equal(true);
```
Comment (jasongrout) [inline tests/test-notebook/src/actions.spec.ts]: This only checks to see if the cell below the current is selected. Can you check all the way to the last cell to make sure each is selected?

**Comment by bw-space** [comment]
@jasongrout are the implementations correct?

**Comment by jasongrout** [comment]
Yes, the implementation and tests look correct now. Thanks!

I did a bit of investigating into the right keyboard shortcut to use. There are two behaviors to investigate:

* What keyboard shortcut highlights from the middle of a document to the top or bottom?
* Consecutive behavior: When you do both keyboard shortcuts one after another, does it highlight everything ("Everything"), or does it just highlight from the middle to the last extreme ("Last Extreme")?


On a mac laptop, PageUp is `Function UpArrow`, PageDown is `Function DownArrow`, Home is `Function LeftArrow` and End is `Function RightArrow`, so a lot of the keystrokes below actually involve pressing the `fn` key.

| Application | Select to top | Select to bottom | Consecutive behavior |
| --- | --- | --- | 
| macOS Finder | `Shift Option UpArrow` | `Shift Option DownArrow` | Everything |
| macOS TextEdit | `Shift Home` | `Shift End` | Everything |
| macOS MS Word | `Cmd Shift Home` | `Cmd Shift End` | Last Extreme |
| macOS Visual Studio Code | None | None | N/A |
| macOS HTML TextArea in Chrome | `Shift Home`, `Cmd Shift UpArrow` | `Shift End`, `Cmd Shift DownArrow` | Everything |
| macOS HTML TextArea in Safari | `Shift Home`, `Cmd Shift UpArrow` | `Shift End`, `Cmd Shift DownArrow` | Everything |
| macOS HTML TextArea in Chrome | `Shift Home`, `Cmd Shift UpArrow` | `Shift End`, `Cmd Shift DownArrow` | Last Extreme |

**Comment by bw-space** [comment]
I think using the last extreme will be a good option, since selecting the entire document can be accomplished using Ctrl+A.

**Comment by jasongrout** [comment]
Can someone run similar checks on Windows and Linux? I tried to test a mix between character-based selections (various text editors), and object-based selections (file browser, spreadsheet, photos, etc.)

**Comment by jasongrout** [comment]
> I think using the last extreme will be a good option, since selecting the entire document can be accomplished using Ctrl+A.

I agree.

**Comment by bw-space** [comment]
@jasongrout any more changes to be made, before merging?

**Comment by bw-space** [comment]
@jasongrout any more changes to be made, before merging?

**Comment by jasongrout** [comment]
> @jasongrout any more changes to be made, before merging?

Thanks for your patience. I think the only thing left is picking the right keyboard shortcuts. The survey on macOS in https://github.com/jupyterlab/jupyterlab/pull/7177#issuecomment-530398463 indicates that probably Accel-Shift-Up/Down is the right keystroke on a mac. As I mention in https://github.com/jupyterlab/jupyterlab/pull/7177#issuecomment-530409232, I'm hoping someone will have time to experiment a bit on windows/linux to see if the corresponding Ctrl-Shift-Up/Down is natural there compared to common applications.

**Comment by bw-space** [comment]
For Ubuntu(Linux):

| Application | Selection Type | Select to Top | Select to Bottom | Consecutive Behaviour |
| -------------   | ------------------   | ----------------  | ----------------------- | ------------------------------ |
| VS Code | Text | Shift + Home | Shift  + End | Last Extreme|
| LibreOffice Writer | Text | Shift + Home | Shift + End | Last Extreme|
| LibreOffice Calc * | Object | Column: Ctrl + Shift + Up Arrow | Column: Ctrl + Shift + Down Arrow | None| 
| HTML TextArea in Firefox | Text | Shift + Home | Shift + End | Last Extreme

LibreOffice Calc : Shift + Home, Shift + End -> Selects upto the last extreme top of current column and last extreme bottom of current column respectively.

**Comment by jasongrout** [comment]
Thanks. Can you try a file browser list of files? That is a common object list that people will have experience with. Also, do you happen to know of an application that has a gnome listbox to see the behavior there?

**Comment by bw-space** [comment]
I have tried with the gedit's( file->open), file browser list, using Shift + Home and Shift + End selects some items, but not upto the top or bottom respectively, and consecutively it repeats selecting the same number of items before or after the currently selected items respectively. I suppose after searching in nautlius(file browser of Gnome) the results are displayed in a list box, and the results are same as the file browser above.

**Comment by jasongrout** [comment]
We're trying on windows:

| Application | Selection type | Select to top | Consecutive behavior |
| --- | --- | --- | --- | --- |
| Windows Explorer (file manager) | Object | `Shift Home`  | Last Extreme |
| Windows Explorer (file manager) | Object | `Ctrl Shift Home`  | Everything |
| Excel 2016 | Object | `Ctrl Shift UpArrow`  | Last Extreme |
| Anki application | Object | `Shift Home` | Everything |
| Wordpad | Text | `Ctrl Shift Home` | Last Extreme |
| Firefox multiselect, Chrome multiselect, Edge multiselect | object | `Shift Home` | Last Extreme |
| Chrome multiselect | object | `Ctrl Shift Home` | Last Extreme |
| Firefox multiselect | object | `Ctrl Shift Home` | Everything |
| Firefox textarea, Chrome textarea, Edge textarea | text | `Ctrl Shift Home` | Last Extreme |

**Comment by jasongrout** [comment]
@bw-space - based on these survey results, how about we have Last Extreme behavior, with the following two sets of shortcuts:

Shift Home / Shift End
Cmd Shift UpArrow, Cmd Shift DownArrow (and explicitly Cmd so that it is mac-specific)

**Comment by jasongrout** [comment]
In other words, let's do Shift Home / Shift End as keybindings and we can merge this, and later we can do the mac-specific keybindings when we figure out a good way to do it.

File: packages/notebook-extension/schema/tracker.json [hunk 15]
```diff
@@ -87,11 +87,21 @@
       "keys": ["Shift K"],
       "selector": ".jp-Notebook:focus"
     },
+    {
+      "command": "notebook:extend-marked-cells-top",
+      "keys": ["Shift PageUp"],
```
Comment (jasongrout) [inline packages/notebook-extension/schema/tracker.json]:
```suggestion
      "keys": ["Shift Home"],
```

File: packages/notebook-extension/schema/tracker.json [hunk 16]
```diff
@@ -87,11 +87,21 @@
...
+      "selector": ".jp-Notebook:focus"
+    },
     {
       "command": "notebook:extend-marked-cells-below",
       "keys": ["Shift ArrowDown"],
       "selector": ".jp-Notebook:focus"
     },
+    {
+      "command": "notebook:extend-marked-cells-bottom",
+      "keys": ["Shift PageDown"],
```
Comment (jasongrout) [inline packages/notebook-extension/schema/tracker.json]:
```suggestion
      "keys": ["Shift End"],
```

**Comment by jasongrout** [comment]
I suggested the changes to the keybindings above. If those look good to you (given the survey results above), feel free to commit the suggestions, and then I can merge.

Thanks again for working on this, and your patience in the process.

**Comment by bw-space** [comment]
@jasongrout, I am very thankful to you for helping me in the process, you have been one of my constant guides. I am looking forward to solve more of the issues :D . It's feels great to contribute here.

**Comment by jasongrout** [comment]
I put in a PR to make it possible for us to have platform-specific bindings: #7335. So after that is merged, we can introduce the mac-specific binding too.

**Comment by jasongrout** [comment]
One last test - and it works great! Thanks again. I look forward to working with you again on another issue.

**Comment by jasongrout** [comment]
@meeseeksdev backport to 1.x

**Comment by bw-space** [comment]
Thanks a lot, and same here. It would be great to have platform specific key bindings.

**Comment by bw-space** [comment]
@jasongrout there has been a slight error in the key bindings above, and has been corrected.
As with the list views the key bindings are correct.  

> For Ubuntu(Linux):
> 
> Application	Selection Type	Select to Top	Select to Bottom	Consecutive Behaviour
> VS Code	Text	Ctrl + Shift + Home (Shift + Page Up)	Ctrl + Shift + End (Shift + Page Down)	Last Extreme
> LibreOffice Writer	Text	Ctrl + Shift + Home (Shift + Page Up)	Ctrl + Shift + End (Shift + Page Down)	Last Extreme
> LibreOffice Calc *	Object	Column: Ctrl + Shift + Up Arrow	Column: Ctrl + Shift + Down Arrow	None
> HTML TextArea in Firefox	Text	Ctrl + Shift + Home (Shift + Page Up)	Ctrl + Shift + End (Shift + Page Down)	Last Extreme
> LibreOffice Calc : Shift + Home, Shift + End -> Selects upto the last extreme top of current column and last extreme bottom of current column respectively.

**Pull request merged**